### ISSUE TITLE
All/most widgets refresh whenever I open a Youtube video

### ISSUE THREAD

**Issue by fishbiscuit13** [felixhageloh/uebersicht#189]
I don't know what could possibly be causing this. I have 6 monitoring widgets (cpu, disk space, istat, etc), and most times that I open a youtube video in Chrome, whether in a new tab, the same tab, or an embedded one, at least 4 of them will refresh. It's not like I'm out of memory of my cpu is going crazy, they're both relatively minimal.It occasionally occurs with other pages, even just opening a new blank tab, but most consistently with Youtube.

**Comment by felixhageloh** [comment]
Wow, odd! 
Could you check the system logs (Console App) just after this occurs and paste a screenshot here? (It looks something like this)
![screen shot 2016-03-25 at 9 12 44](https://cloud.githubusercontent.com/assets/367503/14040677/c1072146-f269-11e5-8177-bf8e137a9bb0.png)

**Comment by fishbiscuit13** [comment]
Yeah, it keeps happening while I'm doing something else and when I open up the console I can't quite figure out which event correlates with it. I'll just keep it open and see if I can get a match.

**Comment by fishbiscuit13** [comment]
Here we go, got two refreshes within a minute, one from a video and one just from a new tab.
<img width="1473" alt="screen shot 2016-03-25 at 01 38 34" src="https://cloud.githubusercontent.com/assets/5496516/14041040/a640723e-f22a-11e5-8574-272b4040d841.png">
<img width="1473" alt="screen shot 2016-03-25 at 01 39 01" src="https://cloud.githubusercontent.com/assets/5496516/14041041/aa7d5ec0-f22a-11e5-9c88-70089ff1ab83.png">

**Comment by felixhageloh** [comment]
I see a lot of 'received display connect changed' messages, which could be the cause of the refresh. It looks like it causes Übersicht to think it is rendering on a new screen (hence the refresh). As for possible solutions I only found this so far:
https://discussions.apple.com/thread/4999113?start=0&tstart=0

**Comment by felixhageloh** [comment]
Had a quick test and thought about it again and I think it is normal. I get the same flickering/refresh when my graphics card switches (this happens when you start some more graphic intensive task, like watching a video). 
I will see if this can be improved - as I mentioned in the previous comment, the problem, is that the screen identifier changes in this case, so Übersicht thinks it is on a new screen.

**Comment by fishbiscuit13** [comment]
Okay so it's still refreshing, but without the Endpoint error, so the trojan isn't the cause. I don't have my other monitor plugged in at the moment, but having it at some point may have caused the issue. I'll try that fix and see what happens.

**Comment by hawkrives** [comment]
@felixhageloh Is the screen identifier consistent between switches? Could we, say, add some heuristics to identify if a switch is for the same screen and only the graphics card changed? Maybe by caching screen identifiers? 

That might be too error prone.

**Comment by felixhageloh** [comment]
@hawkrives cool - curious to hear your findings!

**Comment by hawkrives** [comment]
well, maybe _this_ weekend! I'll add it to my reminders so I actually remember to look 😉

**Comment by hawkrives** [comment]
Quick question– should the screen identifier be on the `<body>`? I only saw `id="__uebersicht"` there.

**Comment by felixhageloh** [comment]
no, not that I can remember @hawkrives

**Comment by hawkrives** [comment]
It'd be this screenID probably, yeah? https://github.com/felixhageloh/uebersicht/blob/aefe9ce51fc2ff654f1bc40fea8601a686ca0e9d/server/src/render.js#L4

**Comment by felixhageloh** [comment]
yeah it is passed down to the render function here https://github.com/felixhageloh/uebersicht/blob/aefe9ce51fc2ff654f1bc40fea8601a686ca0e9d/server/client.coffee#L36

**Comment by hawkrives** [comment]
Good news! I added this right after [`client.coffee#L18`]( https://github.com/felixhageloh/uebersicht/blob/aefe9ce51fc2ff654f1bc40fea8601a686ca0e9d/server/client.coffee#L18), so the screen ID was attached to `<body>`.

```diff
screenId = Number(window.location.pathname.replace(/\//g, ''))
contentEl = document.getElementById('__uebersicht')
+ contentEl.dataset.screenid = screenId
contentEl.innerHTML = ''
```

Then I copied it, changed to the discrete graphics card, and changed back (I used [gfxCardStatus](https://gfx.io)), and I got the same screen ID when I was back to integrated graphics!

```bash
2077752381  # integrated
69731266    # discrete
2077752381  # integrated
````

In short, the screen identifier does appear to be maintained between switches!

**Comment by felixhageloh** [comment]
good findings, thanks for researching! Maybe I'm a bit slow, but I'm still not sure tho how I could tie those two ids to the same screen. Without knowing that only the graphics card changed, this is looks identical to unplugging one screen, connecting another, and then connecting the original screen again

**Comment by hawkrives** [comment]
hmm, yeah, I'm not sure how this knowledge helps yet either. I'll poke around the macOS docs later, maybe, and see if they have, like, a way to enumerate the logical screens instead of the physical screens (or vice versa or something).
